lib/state.go

```diff
@@ -79,3 +95,67 @@ func (s *State) CloneTags() map[string]string {
 	}
 	return tags
 }
+
+// GetScenarioVUID returns the scenario-specific ID of this VU.
+func (s *State) GetScenarioVUID() (uint64, bool) {
+	id, ok := s.scenarioVUID[s.ScenarioName]
+	return id, ok
+}
+
+// SetScenarioVUID sets the scenario-specific ID for this VU.
+func (s *State) SetScenarioVUID(id uint64) {
+	s.scenarioVUID[s.ScenarioName] = id
+}
+
+// GetScenarioVUIter returns the scenario-specific count of completed iterations
+// for this VU.
+func (s *State) GetScenarioVUIter() int64 {
+	s.scIterMx.RLock()
+	defer s.scIterMx.RUnlock()
+	return s.scenarioVUIter[s.ScenarioName]
+}
+
+// IncrIteration increments all iteration counters for the specific VU with this
+// State.
+func (s *State) IncrIteration() {
+	s.scIterMx.Lock()
+	defer s.scIterMx.Unlock()
+
+	s.iteration++
+	if _, ok := s.scenarioVUIter[s.ScenarioName]; ok {
+		s.scenarioVUIter[s.ScenarioName]++
+	} else {
+		s.scenarioVUIter[s.ScenarioName] = 0
+	}
+	if s.IncrScIter != nil {
+		s.scIter = s.IncrScIter()
+	}
+	if s.IncrScIterGlobal != nil {
+		s.scIterGlobal = s.IncrScIterGlobal()
+	}
+}
```

Comment:
This is also logically racy as there is nothing stopping a goroutine going to `IncrScIter` and then before it gets to IncrScIterGlobal 20 others to get to it which leads to this 
```

INFO[0006] 8 9 6 1                                       source=console
INFO[0006] 10 11 5 1                                     source=console
INFO[0006] 12 13 2 1                                     source=console
INFO[0006] 14 15 5 2                                     source=console
INFO[0006] 18 19 5 3                                     source=console
INFO[0006] 17 18 11 0                                    source=console
INFO[0006] 11 12 1 1                                     source=console
INFO[0006] 16 17 4 1                                     source=console
...
```
from 
```
    var stats = execution.getScenarioStats()
    if (stats.iterationGlobal != stats.iteration) {
    console.log(stats.iterationGlobal, stats.iteration, __VU, __ITER)
    }
```

Reply:
This seems like a bad idea idea either way as it lets anything having access to `lib.State` to update the iteration (which was also possible before to be honest, but I still don't like it)
IMO :
1.  `s.iteration` should be part of  InitializedVU 
2. scIter and globalIter be part of ActivatedVU 

And both to somehow get queried from there, but to be update from their respected places *when* an iteration gets started.

Given that the whole iterations things are mostly a roundabout way to do #1539 and that something like [xk6-segment](https://github.com/MStoykov/xk6-segment) will be way more useful (not requiring to have only 1 item per iteration) I am fine with all the iteration counters to be dropped at this point as they seem to be problematic and we can get  1 PR without any of them and then you to do another with what we can do as long as both you and @na-- agree.

Reply:
Hhmm you're right. I don't know what I was thinking with the `State.scIterMx` mutex, it wouldn't work across VUs :man_facepalming:

I see your point about incrementing iterations from `State` being problematic, since it's part of the context and widely shared. Moving the fields to `InitializedVU`/`ActivatedVU` does make sense.

I'll look into fixing this, but I wouldn't want to drop anything at this point, since we've delayed this feature long enough. Let's discuss it later this week.

Reply:
I am mostly proposing that we scale this back to without the iterations so we can merge it more easily and then you to make another PR ... with just the more ... involved iteration changes

Reply:
@MStoykov Please take a look at 9cc9de4b. There are some wonky things about it still, like the setting of lambdas on `lib.State` that are passed via `VUActivationParams`, but I think this sync issue should be fixed now.

We can only retrieve iterations via `State` now, which I needed to expose to `k6/execution`, but the increment is done internally in `js.ActiveVU`.